### PR TITLE
RLPNC-7384: Add Request Object in Java Bindings

### PR DESCRIPTION
**Pull request by meganlevin**

### CODE REVIEW

File: model/src/main/java/com/basistech/rosette/apimodel/recordsimilarity/RecordSimilarityRequest.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.Value;
+
+import java.util.Map;
+
+@Data
+@AllArgsConstructor
+@Value
+public class RecordSimilarityRequest {
```
Comment (fhasanaj) [inline model/src/main/java/com/basistech/rosette/apimodel/recordsimilarity/RecordSimilarityRequest.java]: needs to extend Request

Comment (nnagyprecognoxcom) [reply model/src/main/java/com/basistech/rosette/apimodel/recordsimilarity/RecordSimilarityRequest.java]: done

File: model/src/main/java/com/basistech/rosette/apimodel/recordsimilarity/records/NameRecord.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+    @NoArgsConstructor
+    @ToString(callSuper = true)
+    @EqualsAndHashCode(callSuper = true)
+    @Getter
+    @Setter
+    @JsonInclude(JsonInclude.Include.NON_NULL)
+    public static class FieldedNameRecord extends NameRecord {
+        private String entityType;
+        private LanguageCode language;
```
Comment (fhasanaj) [inline model/src/main/java/com/basistech/rosette/apimodel/recordsimilarity/records/NameRecord.java]: we also want to add script and languageOfOrigin since we want to support same fields as in rni-es

File: model/src/main/java/com/basistech/rosette/apimodel/recordsimilarity/records/NameRecord.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+        public FieldedNameRecord(final String text, final String entityType, final LanguageCode language) {
```
Comment (gcalianese) [inline model/src/main/java/com/basistech/rosette/apimodel/recordsimilarity/records/NameRecord.java]: We also want to support the various combinations of fields

File: json/src/main/java/com/basistech/rosette/apimodel/jackson/recordsimilaritydeserializers/RecordSimilarityDeserializer.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            final Map<String, RecordSimilarityField> fields = node.get("fields").traverse(jsonParser.getCodec()).readValueAs(new TypeReference<Map<String, RecordSimilarityField>>() { });
+            final RecordSimilarityProperties properties = node.get("properties").traverse(jsonParser.getCodec()).readValueAs(RecordSimilarityProperties.class);
+            final RecordSimilarityRecords records = new RecordSimilarityRecords(
+                    parseRecords(node.get("records").get("left"), fields, jsonParser),
+                    parseRecords(node.get("records").get("right"), fields, jsonParser));
+            return new RecordSimilarityRequest(null, fields, properties, records);
+        }
+    }
+
+    private static List<Map<String, Record>> parseRecords(final JsonNode arrayNode,
```
Comment (gcalianese) [inline json/src/main/java/com/basistech/rosette/apimodel/jackson/recordsimilaritydeserializers/RecordSimilarityDeserializer.java]: Not sure if this is the correct place, but maybe add a check for the size of the list? We want to be able to limit the number of records

File: model/src/main/java/com/basistech/rosette/apimodel/recordsimilarity/records/NameRecord.java (hunk 3 above)
Comment (meganlevin) [reply model/src/main/java/com/basistech/rosette/apimodel/recordsimilarity/records/NameRecord.java]: it is supported-- that all args constructor is only used for testing. jackson is using the no args constructor and setters to create the java object when deserializing, with the only required field being the text

File: model/src/main/java/com/basistech/rosette/apimodel/recordsimilarity/records/RecordType.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+import com.fasterxml.jackson.annotation.JsonValue;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+
+import java.util.Arrays;
+
+@AllArgsConstructor
+@Getter
+public enum RecordType {
```
Comment (ethteck) [inline model/src/main/java/com/basistech/rosette/apimodel/recordsimilarity/records/RecordType.java]: I would think this should be FieldType (or RecordFieldType if we're worried that sounds too ambiguous), since it's the type of a field, where a record has multiple fields

File: json/src/test/java/com/basistech/rosette/apimodel/ModelTest.java [hunk 6]
```diff
@@ -101,6 +100,10 @@ void packageTest(boolean inputStreams) throws ClassNotFoundException, IOExceptio
                 continue;
             }
 
+            if (className.contains("RecordSimilarityRequest")) {
```
Comment (meganlevin) [inline json/src/test/java/com/basistech/rosette/apimodel/ModelTest.java:102]: why is this needed?

Comment (nnagyprecognoxcom) [reply json/src/test/java/com/basistech/rosette/apimodel/ModelTest.java:102]: this test is failing because of the deserializer returns [] to the left/right fields instead of a null, when the input is actually null. 
I just wanted a quick fix. I dont think this test means a lot to record similarity.

File: json/src/test/java/com/basistech/rosette/apimodel/ModelTest.java [hunk 7]
```diff
@@ -251,12 +250,12 @@ private Object createObjectForType(Class<?> type, Type genericParameterType, boo
 
         case "Collection":
         case "List": {
+            List<Object> list = new ArrayList<>();
```
Comment (Katsuya-Tomioka) [inline json/src/test/java/com/basistech/rosette/apimodel/ModelTest.java]: Empty fields (list, map, set, etc..) are null, like ADM. Let's not change the assumption.

File: model/src/main/java/com/basistech/rosette/apimodel/recordsimilarity/RecordSimilarityExplainInfo.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+import lombok.Value;
+import lombok.extern.jackson.Jacksonized;
+
+@Builder
+@Jacksonized
+@Value
+public class RecordSimilarityExplainInfo {
+    double weight;
+    @JsonProperty(value = "calculated_weight")
```
Comment (Katsuya-Tomioka) [inline model/src/main/java/com/basistech/rosette/apimodel/recordsimilarity/RecordSimilarityExplainInfo.java]: Curious, names with "_" are from a spec?

Comment (meganlevin) [reply model/src/main/java/com/basistech/rosette/apimodel/recordsimilarity/RecordSimilarityExplainInfo.java]: im not totally sure bc i didnt add this, but i believed they are named like this to mirror the rni-es response

Comment (ethteck) [reply model/src/main/java/com/basistech/rosette/apimodel/recordsimilarity/RecordSimilarityExplainInfo.java]: Yeah @Katsuya-Tomioka , these are from our spec doc (I'll message it to you). If you think we should be using camelCase instead, happy to raise that with the team

Comment (ethteck) [reply model/src/main/java/com/basistech/rosette/apimodel/recordsimilarity/RecordSimilarityExplainInfo.java]: Actually, I don't think we explicitly wrote these in our doc. we should probably just stick to camel case everywhere @meganlevin . After that I think this is good to go

Comment (ethteck) [reply model/src/main/java/com/basistech/rosette/apimodel/recordsimilarity/RecordSimilarityExplainInfo.java]: we've decided to use camel case

File: json/src/main/java/com/basistech/rosette/apimodel/jackson/recordsimilaritydeserializers/RecordSimilarityResultDeserializer.java [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+public class RecordSimilarityResultDeserializer extends StdDeserializer<RecordSimilarityResult> {
+
+    public RecordSimilarityResultDeserializer() {
+        super(RecordSimilarityResult.class);
+    }
+
+    @Override
+    public RecordSimilarityResult deserialize(JsonParser jsonParser, DeserializationContext deserializationContext) throws IOException {
+        final JsonNode node = jsonParser.getCodec().readTree(jsonParser);
+        final Map<String, RecordSimilarityFieldInfo> fields = node.get("fields").traverse(jsonParser.getCodec()).readValueAs(new TypeReference<Map<String, RecordSimilarityFieldInfo>>() { });
```
Comment (nnagyprecognoxcom) [inline json/src/main/java/com/basistech/rosette/apimodel/jackson/recordsimilaritydeserializers/RecordSimilarityResultDeserializer.java]: result json does NOT have "fields" field. We ha to fix this. Thats why my IT fails.

Comment (meganlevin) [reply json/src/main/java/com/basistech/rosette/apimodel/jackson/recordsimilaritydeserializers/RecordSimilarityResultDeserializer.java]: im not sure i understand what you mean. in `RecordSimilarityRequestTest` it should be testing that it serializes and deserializes correctly, including having the "fields" field

Comment (meganlevin) [reply json/src/main/java/com/basistech/rosette/apimodel/jackson/recordsimilaritydeserializers/RecordSimilarityResultDeserializer.java]: oh! sorry just realized this is about the result object, not the request one. yes there was a change to the spec and now it should no longer include the fields object.

Comment (nnagyprecognoxcom) [reply json/src/main/java/com/basistech/rosette/apimodel/jackson/recordsimilaritydeserializers/RecordSimilarityResultDeserializer.java]: This method deserializes json String to a RecordSimialrityResult object. The problem is that we expect that the json text has a "fields" field (at line 41) which the result from the query doesnt have. For example here is an example result:
`{"extendedInformation":null,"results":[{"score":0.9155403504970899,"left":{"name1":"John Malecki"},"right":{"name1":"Malecki John"}}],"errorMessage":""}`
Yesterday i spoke with @fhasanaj about this and she said that this method is not going to be used. 
But I disaggrie: We (currently) are not able to convert the result json to a RecordSimilarityRespone object. Think to the java binding examples. I think in that code we want to deserialize back the json back to a previously mentioned object.

File: model/src/main/java/com/basistech/rosette/apimodel/Request.java [hunk 10]
```diff
@@ -17,17 +17,21 @@
...
 import lombok.AllArgsConstructor;
-import lombok.Data;
+import lombok.Value;
+import lombok.experimental.NonFinal;
+import lombok.experimental.SuperBuilder;
 
 /**
  * Common base class of all requests.
  */
-@Data
```
Comment (Katsuya-Tomioka) [inline model/src/main/java/com/basistech/rosette/apimodel/Request.java]: Can you explain why this needed?

File: model/src/main/java/com/basistech/rosette/apimodel/Request.java [hunk 11]
```diff
@@ -17,17 +17,21 @@
...
+@NonFinal
+@SuperBuilder
 @AllArgsConstructor
 public abstract class Request {
 
     /**
      * @return profile ID
      */
-    protected String profileId;
+    String profileId;
```
Comment (Katsuya-Tomioka) [inline model/src/main/java/com/basistech/rosette/apimodel/Request.java]: Why? Some test requirement?

File: model/src/main/java/com/basistech/rosette/apimodel/recordsimilarity/RecordSimilarityResponse.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ * Response data model for comparison of two records.
+ */
+@Value
+@Builder
+@Jacksonized
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class RecordSimilarityResponse extends Response {
+    Map<String, RecordSimilarityFieldInfo> fields;
```
Comment (Katsuya-Tomioka) [inline model/src/main/java/com/basistech/rosette/apimodel/recordsimilarity/RecordSimilarityResponse.java]: Sonar says you need "@Valid" here. 
https://sonar.basistech.net/project/issues?id=com.basistech.rosette%3Arosette-api-java-binding&open=AY5SZyAcDn-JuiNfWcaa&pullRequest=215&resolved=false&sinceLeakPeriod=true&types=CODE_SMELL

File: model/src/main/java/com/basistech/rosette/apimodel/recordsimilarity/RecordSimilarityRequest.java [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+import javax.validation.constraints.NotEmpty;
+import javax.validation.constraints.NotNull;
+import java.util.Map;
+
+@Value
+@SuperBuilder
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class RecordSimilarityRequest extends Request {
+    @NotEmpty @Valid Map<String, RecordSimilarityFieldInfo> fields;
+    RecordSimilarityProperties properties;
```
Comment (Katsuya-Tomioka) [inline model/src/main/java/com/basistech/rosette/apimodel/recordsimilarity/RecordSimilarityRequest.java]: Sonar says you need "@Valid" here. 
https://sonar.basistech.net/project/issues?id=com.basistech.rosette%3Arosette-api-java-binding&open=AY5SZyAhDn-JuiNfWcab&pullRequest=215&resolved=false&sinceLeakPeriod=true&types=CODE_SMELL

**Review by Katsuya-Tomioka — COMMENTED** [review]
I left a couple of comments. Besides sonar's about "Valid" annotations, I want to make sure Request is behaving the same. As far as I can tell from jenkins the PR looks ok.

File: model/src/main/java/com/basistech/rosette/apimodel/Request.java (hunk 10 above)
Comment (nnagyprecognoxcom) [reply model/src/main/java/com/basistech/rosette/apimodel/Request.java]: A class beeing immutable is way better than beeing just a record, for example in multithreded environments.
@Value makes the class immutable

File: model/src/main/java/com/basistech/rosette/apimodel/Request.java (hunk 11 above)
Comment (nnagyprecognoxcom) [reply model/src/main/java/com/basistech/rosette/apimodel/Request.java]: @Value makes all the fields private final. No need to annotate them otherwise

Comment (Katsuya-Tomioka) [reply model/src/main/java/com/basistech/rosette/apimodel/Request.java]: But `@NonFinal` was also added to the class. So the intention is to have a private non-final field?

File: model/src/main/java/com/basistech/rosette/apimodel/Request.java [hunk 14]
```diff
@@ -17,17 +17,21 @@
...
+import lombok.experimental.NonFinal;
+import lombok.experimental.SuperBuilder;
 
 /**
  * Common base class of all requests.
  */
-@Data
+@Value
+@NonFinal
+@SuperBuilder
```
Comment (Katsuya-Tomioka) [inline model/src/main/java/com/basistech/rosette/apimodel/Request.java]: I noticed SuperBuilder was added to the Names request side of the hierarchy. The doc says it's not compatible with Builder, which DocRequest uses. What part not compatible do you know? Seems like if we should be using SuperBuilder everywhere.

File: model/src/main/java/com/basistech/rosette/apimodel/Request.java (hunk 10 above)
Comment (Katsuya-Tomioka) [reply model/src/main/java/com/basistech/rosette/apimodel/Request.java]: `NonFinal` makes it mutable. Can you see if you can take it off? That's probably better.

**Comment by Katsuya-Tomioka** [comment]
I think for me the last bit about the PR is SuperBuilder said incompatible with Builder in the doc. I'm not sure how incompatible they are. Can you guys spend little time on it to see? Or maybe DocRequest can be SuperBuilder as well, so we don't use Builder?

File: model/src/main/java/com/basistech/rosette/apimodel/Request.java (hunk 14 above)
Comment (Katsuya-Tomioka) [reply model/src/main/java/com/basistech/rosette/apimodel/Request.java]: I take back, option objects do use `Builder`, but not `DocRequest` or any other request objects.

File: model/src/main/java/com/basistech/rosette/apimodel/Request.java (hunk 11 above)
Comment (nnagyprecognoxcom) [reply model/src/main/java/com/basistech/rosette/apimodel/Request.java]: @NonFinal is only for the class not for the members. @Value makes the class final as well, but we dont want an abstract class being final (not even possible).

File: model/src/main/java/com/basistech/rosette/apimodel/Request.java (hunk 10 above)
Comment (nnagyprecognoxcom) [reply model/src/main/java/com/basistech/rosette/apimodel/Request.java]: Nope, nonfinal makes the class not final not the members.

File: model/src/main/java/com/basistech/rosette/apimodel/Request.java (hunk 11 above)
Comment (Katsuya-Tomioka) [reply model/src/main/java/com/basistech/rosette/apimodel/Request.java]: Got it, so `@Value` here makes pretty confusing to me. Single field abstract class. Let's not use it, just put a getter, keep the constructor.

**Review by Katsuya-Tomioka — CHANGES_REQUESTED** [review]
Please address the comment on Request. After done, please linearlize commits
 and run following Jenkins jobs manually:
http://jenkins-jugg.basistech.net/job/Rosette%20Bindings/job/java/
http://jenkins-jugg.basistech.net/job/RosetteApi/job/java/

For sonar, please fix any critical smells at minimum.
https://sonar.basistech.net/dashboard?id=com.basistech.rosette%3Arosette-api-java-binding&pullRequest=215 

Looks good to me otherwise.

**Review by Katsuya-Tomioka — CHANGES_REQUESTED** [review]
Please double check profileId works for DocumentRequest.
Switching from `@Data` means `Request` lost a setter for profileId. This would mean a breaking change for customers, which we can't have it.
Note names requests don't use profileId. Only document requests use it.

**Comment by iredpath** [comment]
> Please address the comment on Request. After done, please linearlize commits

@Katsuya-Tomioka do you just want things squashed into a single commit, or did you have something else in mind?

**Comment by Katsuya-Tomioka** [comment]
> > Please address the comment on Request. After done, please linearlize commits
> 
> @Katsuya-Tomioka do you just want things squashed into a single commit, or did you have something else in mind?

Squash without preserving comments is fine. Can you add a summary comment instead?

**Review by Katsuya-Tomioka — APPROVED** [review]
